### ISSUE TITLE
Build Addons for Windows

### ISSUE THREAD

**Issue by seanpmorgan** [tensorflow/addons#173]
Splitting this as a separate issue from #77. MacOS should be an achievable goal prior to the 0.3 release. Windows will take a bit longer.

**Comment by geometrikal** [comment]
Any progress?

**Comment by geometrikal** [comment]
To explain why this is important for Windows users: 

If you want to do rotation augmentation, options are either a python library (e.g. numpy) or image rotate from tensorflow addons. Multiprocessing doesn't really work for Windows, e.g. using keras fit_generator with multiprocessing enable will throw an error, so they best (and fastest) way has been to use tf.data.Dataset with mapping via tf.contrib.image.rotate. With TF2.0 contrib has been removed, so until addons works there is no easy way to do parallel rotation augmentation.

**Comment by seanpmorgan** [comment]
Hi @geometrikal! Thanks for bumping this and explaining the need. There is no progress that I am aware of though the issue is tagged `help wanted` so perhaps someone else could expedite it. 

After we finish the nightly builds I can start working on this issue but it may not be for another month or so. It will certainly be added though as this is high priority per our monthly newsletter.

**Comment by svobora** [comment]
I have the same issue as @geometrikal, as I use `tf.data.Dataset`, `map `for preprocessing, there is no workaround for data augmentation with rotations. Please look into it.

**Comment by geometrikal** [comment]
@seanpmorgen The see this has the help-wanted tag. What would be the overall steps required to try to get this running on Windows?

**Comment by seanpmorgan** [comment]
@geometrikal I've created an issue in custom-op repo to discuss the current issue and see how we can get this done.

**Comment by sbarman-mi9** [comment]
Any update on Windows or Universal binaries for this package? It is much appreciated to target Windows users as well as *NIX systems.

**Comment by seanpmorgan** [comment]
Hi @sbarman-mi9 we're currently blocked by this issue: https://github.com/tensorflow/custom-op/issues/24

There will need to be some changes on core TensorFlow's side. We'll bring it back up again at this month's meeting to see if we can get some progress. CC @perfinion to see if he has any more information from SIG-Build

**Comment by davideboschetto** [comment]
Is there any update to this? 
tfa seems quite useful for windows users.....

**Comment by seanpmorgan** [comment]
Hi @davideboschetto unfortunately this is blocked until progress is made in the issue linked above:
https://github.com/tensorflow/custom-op/issues/24

We'll update this as information is made available.

**Comment by gigadeplex** [comment]
Hi I also ran into this problem just now!!

**Comment by consi95** [comment]
Hello :) i know that this is blocked, but would be really nice if there would be any progress. I am stucked at the same point. I dont want to rewrite my code to older tf versions...

**Comment by ChenYang-ChenYang** [comment]
The same request there. I'm always waiting from May to now!

**Comment by meysamsadeghi** [comment]
I also would like to ask for further action on this issue. It is very useful. Also, is there any workaround suggestion?

**Comment by seanpmorgan** [comment]
Hi @meysamsadeghi (and others). We're really sorry for this not being implemented yet, but it's  outside of the scope of what SIG-Addons is capable of supporting at the moment. 

Please forward your request comments/support to the custom-op issue: https://github.com/tensorflow/custom-op/issues/24 so that it can be prioritized. The roadmap has been laid out, but it needs to be completed.

There are a couple of possible work arounds that I see though:

1. Use [Windows Subsystem for Linux (WSL)](https://docs.microsoft.com/en-us/windows/wsl/install-win10) 
2. As a last resort, you could copy addons operations into a fork of tensorflow/tensorflow core repo and compile the custom-ops alongside the entire framework.

**Comment by jkyl** [comment]
Unfortunately WSL does not enable GPU passthrough. I've bumped https://github.com/tensorflow/custom-op/issues/24 with a plea for increased priority.

**Comment by seanpmorgan** [comment]
Thanks to @yifeif this should be unblocked.

**Comment by FaraiCH** [comment]
Hi, I'm using Tensorflow 2.0 and I have come to a problem. tf.contrib is not recognized as an attribute and when I try to install tensorflow-addons on my windows machine, it fails. However, it seems to work fine on my Mac. Is there any news or update on this?

**Comment by geometrikal** [comment]
@FaraiCH As @seanpmorgan has posted, this is waiting on implementation of custom op in the normal Tensorflow repo. It seems there has been some work done on this so we can be hopeful of some progress soon!

**Comment by rawhazard** [comment]
Hi. Is it possible to install TF-Addons on Windows 10 now with pip? If not with pip, is it possible to install it from sources in other ways?
Thank you for your help.

**Comment by seanpmorgan** [comment]
> Hi. Is it possible to install TF-Addons on Windows 10 now with pip? If not with pip, is it possible to install it from sources in other ways?
> Thank you for your help.

Hi @rawhazard not yet, but we're hoping to have a windows CPU package in the very near future

**Comment by drozzy** [comment]
Please add support for windows. I get cryptic error when trying to install this:

```
ERROR: Could not find a version that satisfies the requirement tensorflow-addons (from versions: none)
ERROR: No matching distribution found for tensorflow-addons
```

**Comment by mihaimaruseac** [comment]
@drozzy when you get that message you can get more information by trying `pip debug verbose` and looking at the tags `pip` gives back to you and comparing with the tags on the pypi page. This works for any python package.

**Comment by seanpmorgan** [comment]
PR has been merged, expect a windows wheel available for 0.7 release which is waiting on TF2.1. We can't make nightlies for windows as there are no tf-nightly windows builds.

**Comment by guillaumekln** [comment]
Great news! Thank you very much for your work, Sean!

**Comment by tejasdixit** [comment]
So i guess i will need to do it again with tensorflow 2.1 (currently i am using version 1.5 as the file generate_tfrecord.py won't even work with tensorflow 2.0)

**Comment by seanpmorgan** [comment]
FYI stable release for windows is out. `pip install tensorflow-addons`

**Comment by tejasdixit** [comment]
Thank you for your response.

Regards.

**Comment by yifeif** [comment]
Thank you @seanpmorgan!
